### PR TITLE
doc: guides: introduce ARM Cortex-M user guide

### LINKED ISSUE #36388: ARM: Architecture Level user guide
**Is your enhancement proposal related to a problem? Please describe.**
Generate an architecture-level user guide for ARM Cortex-m architecture variant, similar to the ARC and x86 user guides in the project documentation.

### PR DESCRIPTION
**Pull request by ioannisg**
ARM Cortex-M user guide. Initial version.
- status of the porting
- details on feature support
- important considerations

ToDo in a follow-up PR: TrustZone-M considerations. This will be addressed as part of the TFM guide ticket.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Closes #36388

### CODE REVIEW

File: doc/guides/arch/arm.rst [hunk 1]
```diff
@@ -0,0 +1,79 @@
+.. _arm_developer_guide:
+
+ARM Cortex-M Developer Guide
```
Comment (stephanosio) [inline doc/guides/arch/arm.rst:3]: Noting that this file is named `arm.rst`, the page title should probably be something like `ARM Developer Guide` with the Cortex-M-specific details in a subsection.

File: doc/guides/arch/arm.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+|                                 |   Direct  interrupts              |   Y    |    Y   |   Y     |   Y    |    Y      |    Y   |    Y    |     Y      |   Y        |
++---------------------------------+-----------------------------------+--------+--------+---------+--------+-----------+--------+---------+------------+------------+
+|                                 |   Zero Latency interrupts         |   N    |    N   |   N     |   Y    |    Y      |    Y   |    Y    |     Y      |   Y        |
++---------------------------------+-----------------------------------+--------+--------+---------+--------+-----------+--------+---------+------------+------------+
+| CPU idling                      |                                   |   Y    |    Y   |   Y     |   Y    |    Y      |    Y   |    Y    |     Y      |   Y        |
++---------------------------------+-----------------------------------+--------+--------+---------+--------+-----------+--------+---------+------------+------------+
+| Native system timer (SysTick)   |                                   |   N    |    N   |   Y     |   Y    |    Y      |    Y   |    Y    |     Y      |   Y        |
```
Comment (henrikbrixandersen) [inline doc/guides/arch/arm.rst:47]: The Cortex-M1 IP has optional SysTick support. This is supported by Zephyr. Perhaps we need to expand the `Y`/`N` with `O` for optional?

Comment (ioannisg) [reply doc/guides/arch/arm.rst:47]: I was thinking about it, yes.
I think we need either some increased legend or some list of footnotes..

File: doc/guides/arch/arm.rst (hunk 1 above)
Comment (ioannisg) [reply doc/guides/arch/arm.rst:3]: I am not ready to work on the Cortex-R, though @stephanosio , so I wonder if i should just stick to the cortex-m user guide. I can rename the file to arm_cortex_m.rst, to make this more clear. Once we are ready to add a cortex-r guide, we can do file renaming

Comment (ioannisg) [reply doc/guides/arch/arm.rst]: Renamed this to arm_cortex_m.rst

File: doc/guides/arch/arm.rst (hunk 2 above)
Comment (ioannisg) [reply doc/guides/arch/arm.rst]: Done, added footnote @henrikbrixandersen 
(this is how the ARC guys have populated their arch-support-status table: using foot notes :) )

File: doc/guides/arch/arm_cortex_m.rst [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+***********
+
+Threads
+=======
+
+Thread stack alignment
+----------------------
+
+Each Zephyr thread is defined with its own stack memory. By default, Cortex-M enforces a double word thread stack alignment, see
+:option:`CONFIG_STACK_ALIGN_DOUBLE_WORD`. If MPU-based HW-assisted stack overflow detection (:option:`CONFIG_MPU_STACK_GUARD`)
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:100]: All Kconfig options need to be referenced now using the `:kconfig:` role

**Review by gmarull — CHANGES_REQUESTED** [review]

File: doc/guides/arch/arm_cortex_m.rst (hunk 3 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: Fixed that now @gmarull , thanks.

File: doc/guides/arch/arm_cortex_m.rst [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+------------------
+
+Additional *static* MPU regions are programmed once during system boot. These regions
+are required to enable certain features
+
+* a RX region to allow execution from SRAM, when :kconfig:`CONFIG_ARCH_HAS_RAMFUNC_SUPPORT` is
+  enabled and users have defined functions to execute from SRAM.
+* a RX region for relocating text sections to SRAM when :kconfig:`CONFIG_CODE_DATA_RELOCATION_SRAM` is enabled
+* a no-cache region to allow for a none-cacheable SRAM area when :kconfig:`CONFIG_NOCACHE_MEMORY` is enabled
+* a possibly unprivileged RW region for GCOV code coverage accouning area, when :kconfig:`CONFIG_COVERAGE_GCOV` is enabled
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:443]:
```suggestion
* a possibly unprivileged RW region for GCOV code coverage accounting area, when :kconfig:`CONFIG_COVERAGE_GCOV` is enabled
```

File: doc/guides/arch/arm_cortex_m.rst [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+developing Zephyr applications for Arm Cortex-M-based platforms.
+
+Key supported features
+**********************
+
+The table below summarizes the status of key OS features in the different
+Arm Cortex-M implementation variants.
+
+
++---------------------------------+-----------------------------------+-----------------+---------+--------+-----------+--------+---------+------------+------------+
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:20]: `.. list-table::` may lead to more readable table, see e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/contribute/coding_guidelines/index.rst

File: doc/guides/arch/arm_cortex_m.rst [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+* When switching-out the current thread, the processor stores
+
+   * the callee-saved registers (R4 - R11) in the thread's container for callee-saved registers,
+     which is located in kernel memory
+   * the thread's current operation *mode*
+
+        * user or privileged execution mode
+        * presense of an active floating point context
+        * the EXC_RETURN value of the current handler context (PendSV)
+   * the floating point callee-saved registers (S16 - S31) in the thread's container for FP
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:144]:
```suggestion
        * the EXC_RETURN value of the current handler context (PendSV)

   * the floating point callee-saved registers (S16 - S31) in the thread's container for FP
```

File: doc/guides/arch/arm_cortex_m.rst [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+
+Interrupt priority levels
+-------------------------
+
+The number of available (configurable) interrupt priority levels is
+determined by the number of implemented interrupt priority bits in
+NVIC; this needs to be described for each Cortex-M platform using
+DeviceTree:
+
+   .. code-block:: c
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:206]:
```suggestion
.. code-block:: c
```

File: doc/guides/arch/arm_cortex_m.rst [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+* Regular HW interrupts are assigned priority levels lower than SVC.
+
+The priority level configuration in Cortex-M is implemented in
+:file:`include/arch/arm/aarch32/exc.h`.
+
+Locking and unlocking IRQs
+--------------------------
+In Baseline Cortex-M locking interrupts is implemented using the PRIMASK register.
+
+   .. code-block:: c
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:267]:
```suggestion
.. code-block:: c
```

File: doc/guides/arch/arm_cortex_m.rst [hunk 9]
```diff
@@ -0,0 +1,643 @@
...
+* ZLIs are assigned the highest configurable priority level
+* SVCs are assigned the second highest configurable priority level
+* Regular HW interrupts are assigned priority levels lower than SVC.
+
+The priority level configuration in Cortex-M is implemented in
+:file:`include/arch/arm/aarch32/exc.h`.
+
+Locking and unlocking IRQs
+--------------------------
+In Baseline Cortex-M locking interrupts is implemented using the PRIMASK register.
```
Comment (gmarull) [inline doc/guides/arch/arm_cortex_m.rst:265]:
```suggestion
--------------------------

In Baseline Cortex-M locking interrupts is implemented using the PRIMASK register.
```

File: doc/guides/arch/arm_cortex_m.rst (hunk 7 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: Yeap, just realized this is not c code :) 
This was also leading to a build failure. I have changed to console, but perhaps devicetree is a better match

File: doc/guides/arch/arm_cortex_m.rst [hunk 10]
```diff
@@ -0,0 +1,643 @@
...
+| dereference detection           |                                   |        N        |   N     |   Y    |    Y      |    Y   |    Y    |     Y      |   Y        |
++---------------------------------+-----------------------------------+-----------------+---------+--------+-----------+--------+---------+------------+------------+
+| HW-assisted atomic operations   |                                   |        N        |   N     |   Y    |    Y      |    Y   |    N    |     Y      |   Y        |
++---------------------------------+-----------------------------------+-----------------+---------+--------+-----------+--------+---------+------------+------------+
+|Support for non-cacheable regions|                                   |        N        |   N     |   Y    |    Y      |    Y   |    N    |     Y      |   Y        |
++---------------------------------+-----------------------------------+-----------------+---------+--------+-----------+--------+---------+------------+------------+
+| Execute from SRAM               |                                   |        N        |   N     |   Y    |    Y      |    Y   |    N    |     Y      |   Y        |
```
Comment (henrikbrixandersen) [inline doc/guides/arch/arm_cortex_m.rst:65]: The Cortex-M1 normally executes code from SRAM/Block RAM within an FPGA. Perhaps another footnote is needed?

**Review by henrikbrixandersen — APPROVED** [review]
Thank you for this guide. It is very well-written and worth a read for anybody developing Zephyr applications for Arm Cortex-M.

File: doc/guides/arch/arm_cortex_m.rst (hunk 4 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: Fixed

File: doc/guides/arch/arm_cortex_m.rst (hunk 6 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: done

File: doc/guides/arch/arm_cortex_m.rst (hunk 7 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: done now

File: doc/guides/arch/arm_cortex_m.rst (hunk 8 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: done

File: doc/guides/arch/arm_cortex_m.rst (hunk 9 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: done

File: doc/guides/arch/arm_cortex_m.rst (hunk 5 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst:20]: The problem with this table design is that it is not possible to have mixed cell configuration, e.g. merged rows and columns arbitrarily (or at least I have not found an easy way to do it).
The tables in this guide use heavily rows and columns of variable sizes, so I am a bit reluctant to switch to using a more restrictive table framework.

File: doc/guides/arch/arm_cortex_m.rst (hunk 10 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst:65]: Right, this is more of a feature on XIP images to execute from SRAM, when code is normally in flash. I'll rephrase this, thanks.

**Review by gmarull — APPROVED** [review]
LGTM, great page!

File: doc/guides/arch/arm_cortex_m.rst (hunk 10 above)
Comment (ioannisg) [reply doc/guides/arch/arm_cortex_m.rst]: fixed.

**Pull request merged**